K8s object cache altered at runtime

```

CACHE *resource.resource[*github.com/cilium/cilium/pkg/k8s/slim/k8s/api/networking/v1.NetworkPolicy][0] ALTERED!
(*v1.NetworkPolicy)({                                                                        (*v1.NetworkPolicy)({
  TypeMeta: (v1.TypeMeta) {                                                                    TypeMeta: (v1.TypeMeta) {
    Kind: (string) "",                                                                           Kind: (string) "",
    APIVersion: (string) ""                                                                      APIVersion: (string) ""
  },                                                                                           },
  ObjectMeta: (v1.ObjectMeta) {                                                                ObjectMeta: (v1.ObjectMeta) {
    Name: (string) (len=31) "allow-client-a-via-pod-selector",                                   Name: (string) (len=31) "allow-client-a-via-pod-selector",
    GenerateName: (string) "",                                                                   GenerateName: (string) "",
    Namespace: (string) (len=1) "x",                                                             Namespace: (string) (len=1) "x",
    UID: (types.UID) (len=36) "cf23c67b-9f58-4452-b179-2ef152b5cd60",                            UID: (types.UID) (len=36) "cf23c67b-9f58-4452-b179-2ef152b5cd60",
    ResourceVersion: (string) (len=4) "1849",                                                    ResourceVersion: (string) (len=4) "1849",
    Generation: (int64) 1,                                                                       Generation: (int64) 1,
    DeletionTimestamp: (*v1.Time)(<nil>),                                                        DeletionTimestamp: (*v1.Time)(<nil>),
    Labels: (map[string]string) <nil>,                                                           Labels: (map[string]string) <nil>,
    Annotations: (map[string]string) <nil>,                                                      Annotations: (map[string]string) <nil>,
    OwnerReferences: ([]v1.OwnerReference) <nil>                                                 OwnerReferences: ([]v1.OwnerReference) <nil>
  },                                                                                           },
  Spec: (v1.NetworkPolicySpec) {                                                               Spec: (v1.NetworkPolicySpec) {
    PodSelector: (v1.LabelSelector) {                                                            PodSelector: (v1.LabelSelector) {
      MatchLabels: (map[string]string) (len=1) {                                                   MatchLabels: (map[string]string) (len=1) {
        (string) (len=3) "pod": (string) (len=1) "a"                                                 (string) (len=3) "pod": (string) (len=1) "a"
      },                                                                                           },
      MatchExpressions: ([]v1.LabelSelectorRequirement) <nil>                                      MatchExpressions: ([]v1.LabelSelectorRequirement) <nil>
    },                                                                                           },
    Ingress: ([]v1.NetworkPolicyIngressRule) (len=1) {                                           Ingress: ([]v1.NetworkPolicyIngressRule) (len=1) {
      (v1.NetworkPolicyIngressRule) {                                                              (v1.NetworkPolicyIngressRule) {
        Ports: ([]v1.NetworkPolicyPort) <nil>,                                                       Ports: ([]v1.NetworkPolicyPort) <nil>,
        From: ([]v1.NetworkPolicyPeer) (len=1) {                                                     From: ([]v1.NetworkPolicyPeer) (len=1) {
          (v1.NetworkPolicyPeer) {                                                                     (v1.NetworkPolicyPeer) {
            PodSelector: (*v1.LabelSelector)({                                                           PodSelector: (*v1.LabelSelector)({
              MatchLabels: (map[string]string) (len=3) {                                                   MatchLabels: (map[string]string) (len=2) {
                (string) (len=3) "pod": (string) (len=1) "b",                                                (string) (len=3) "pod": (string) (len=1) "b",
                (string) (len=4) "pod2": (string) (len=7) "updated",                                         (string) (len=4) "pod2": (string) (len=7) "updated"
                (string) (len=28) "io.cilium.k8s.policy.cluster": (string) (len=7) "default"               },
              },                                                                                           MatchExpressions: ([]v1.LabelSelectorRequirement) <nil>
              MatchExpressions: ([]v1.LabelSelectorRequirement) <nil>                                    }),
            }),                                                                                          NamespaceSelector: (*v1.LabelSelector)(<nil>),
            NamespaceSelector: (*v1.LabelSelector)(<nil>),                                               IPBlock: (*v1.IPBlock)(<nil>)
            IPBlock: (*v1.IPBlock)(<nil>)                                                              }
          }                                                                                          }
        }                                                                                          }
      }                                                                                          },
    },                                                                                           Egress: ([]v1.NetworkPolicyEgressRule) <nil>,
    Egress: ([]v1.NetworkPolicyEgressRule) <nil>,                                                PolicyTypes: ([]v1.PolicyType) (len=1) {
    PolicyTypes: ([]v1.PolicyType) (len=1) {                                                       (v1.PolicyType) (len=7) "Ingress"
      (v1.PolicyType) (len=7) "Ingress"                                                          }
    }                                                                                          }
  }                                                                                          })
})                                                                                           
   
panic: cache *resource.resource[*github.com/cilium/cilium/pkg/k8s/slim/k8s/api/networking/v1.NetworkPolicy] modified

goroutine 843 [running]:
k8s.io/client-go/tools/cache.(*defaultCacheMutationDetector).CompareObjects(0xc00348fb80)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/mutation_detector.go:165 +0x58e
k8s.io/client-go/tools/cache.(*defaultCacheMutationDetector).Run(0xc00348fb80, 0xc001261110)
	/go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/mutation_detector.go:109 +0x149
created by github.com/cilium/cilium/pkg/k8s/resource.(*wrapperController).Run in goroutine 830
	/go/src/github.com/cilium/cilium/pkg/k8s/resource/resource.go:732 +0x87
                                                                                          
```

The problem is this line https://github.com/cilium/cilium/blob/461100f9e5f4986c1e82045882a52e87918541b2/pkg/k8s/network_policy.go#L92

and it might have been added in https://github.com/cilium/cilium/commit/e80eea9f06aa5f08f233a5897d3acf31f130edf5. cc @MrFreezeex @giorio94 

We had a similar issue as this one in the past and it got fixed with this commit https://github.com/cilium/cilium/commit/042e70329e137377d4887394d789dc2f1ba5ead7

Perhaps we can reuse the same solution as that one.

--- a/pkg/k8s/network_policy.go
+++ b/pkg/k8s/network_policy.go
@@ -79,17 +79,21 @@ func parseNetworkPolicyPeer(clusterName, namespace string, peer *slim_networking
 		return nil
 	}
 
+	// peer should not be nutated in this function
+
 	var retSel *api.EndpointSelector
+	podSelector := peer.PodSelector.DeepCopy()
+
 	// The PodSelector should only reflect to the configured cluster unless the selector
 	// explicitly targets another cluster already.
-	if clusterName != cmtypes.PolicyAnyCluster && !isPodSelectorSelectingCluster(peer.PodSelector) {
-		if peer.PodSelector == nil {
-			peer.PodSelector = &slim_metav1.LabelSelector{}
+	if clusterName != cmtypes.PolicyAnyCluster && !isPodSelectorSelectingCluster(podSelector) {
+		if podSelector == nil {
+			podSelector = &slim_metav1.LabelSelector{}
 		}
-		if peer.PodSelector.MatchLabels == nil {
-			peer.PodSelector.MatchLabels = map[string]slim_metav1.MatchLabelsValue{}
+		if podSelector.MatchLabels == nil {
+			podSelector.MatchLabels = map[string]slim_metav1.MatchLabelsValue{}
 		}
-		peer.PodSelector.MatchLabels[k8sConst.PolicyLabelCluster] = clusterName
+		podSelector.MatchLabels[k8sConst.PolicyLabelCluster] = clusterName
 	}
 
 	if peer.NamespaceSelector != nil {
@@ -123,10 +127,10 @@ func parseNetworkPolicyPeer(clusterName, namespace string, peer *slim_networking
 			namespaceSelector.MatchExpressions = []slim_metav1.LabelSelectorRequirement{allowAllNamespacesRequirement}
 		}
 
-		selector := api.NewESFromK8sLabelSelector(labels.LabelSourceK8sKeyPrefix, namespaceSelector, peer.PodSelector)
+		selector := api.NewESFromK8sLabelSelector(labels.LabelSourceK8sKeyPrefix, namespaceSelector, podSelector)
 		retSel = &selector
-	} else if peer.PodSelector != nil {
-		podSelector := parsePodSelector(peer.PodSelector, namespace)
+	} else if podSelector != nil {
+		podSelector = parsePodSelector(podSelector, namespace)
 		selector := api.NewESFromK8sLabelSelector(labels.LabelSourceK8sKeyPrefix, podSelector)
 		retSel = &selector
 	}

